[Yuen Si Hao] iP

yuen-sihao opened the pull request.

Commit:
Add Increment: Level 0 - Greet

A skeletal version of the Duke that simply greets the user and exits.

Commit:
Add ability to echo user commands

A skeletal version of Duke that greets user, echos commands entered by
the user, and exits when the user types "bye".

Commit:
Add ability to store user text and display them

Duke now has the ability to store text entered by the user and display
them when requested.

Commit:
Add ability to mark tasks as done

Duke can now allow user to mark tasks as done.

Commit:
Tweak code to comply with coding standard better

Changes were made to the naming and use of boolean variable.

Commit:
Update code to include getters, setters, and class-level members

Improvements are made to the code to include the use of getters,
setters and class-level members.

Commit:
Add ability to differentiate tasks by type

Duke now has the ability to track three types of tasks such as ToDos,
Deadlines, and Events.

Commit:
Change class name to suit coding standard better

The name of the classes are changed to singular nouns to suit the
coding standard better, causing these files to be removed.

Commit:
Update code to improve the use of inheritance

The code is modified to improve the use of inheritance.

Commit:
Update Code to improve code quality

Some parts of the code are refactored to improve code quality.

Commit:
Revert "Update Code to improve code quality"

This reverts commit d8ba49c335275b3c58efb3d26d3243fd87a88bb1.

Commit:
Update code to improve code quality

Some parts of the code are refactored to improve code quality.

Commit:
Update code to semi-automate testing of Duke

* Add script, testcase and expected outcome for automated testing
* Solve EOF error during automated testing

Commit:
Update code to improve Duke

* Output messages by Duke are customized
* Check for unknown tasks not recognized by Duke

Commit:
Update code to improve code quality

Some parts of the code is refactored to improve code quality.

samuellleow commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,138 @@
+        printDetailsOfAddedTask(newTask);
+    }
+
+    private static void printInvalidTaskMessage() {
+        System.out.println("I'm sorry I don't understand you." + System.lineSeparator()
+                + "Would you like to tell me again?");
+        printSingleLine();
```
Great job on extracting your code into different readable functions with clear function heading.
And also a fantastic job on error handling.

yuen-sihao replied:
Thanks for your comments!

samuellleow reviewed (commented):
Well done!! Your code is looking good.

LiewWS commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,138 @@
+        printSingleLine();
+    }
+
+    private static void updateListOfTask(String command, Task[] tasks) {
```
The code is well written and although variables and method names are clear, maybe adding comments could make the code easier to understand.

yuen-sihao replied:
Ok, will take note on the point for adding comments. Thanks for the comments!

LiewWS reviewed (commented):
Generally very well written.

Commit:
Update code to improve code quality

There are methods that may not be necessary and the code can be
simplified.

Commit:
Update expected outcome of automated test

The expected outcome has changed as the code has been modified.

Commit:
Include the use of exceptions to handle errors

The code has implemented the use of exceptions to handle errors.

Commit:
Merge branch 'branch-Level-5'

Commit:
Divide classes into packages

The classes of code in Duke are now organized into suitable
java packages.

Commit:
Merge branch 'branch-A-Packages'

LiewWS commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,182 @@
+import java.util.Scanner;
+
+public class Duke {
+    public static final int MAX_NUMBER_OF_TASK = 100;
```
Good to see that the correct naming format is being used for constants.

LiewWS commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    private static void determineTypeOfTask(Task[] tasks, String typeOfTask, String description) {
+        switch (typeOfTask) {
+        case "todo":
+            createTodoTask(tasks, description);
+            break;
+        case "deadline":
+            checkValidityOfDeadline(tasks, description);
+            break;
+        case "event":
+            checkValidityOfEvent(tasks, description);
+            break;
+        case "done":
+            checkValidityOfCompletedTask(tasks, description);
+            break;
+        default:
```
Nice alignment of the case labels to the switch statement according to the standard.

LiewWS commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,182 @@
+    public static void main(String[] args) {
+        printWelcomeMessage();
+        String command;
+        Task[] tasks = new Task[MAX_NUMBER_OF_TASK];
```
I like how you place the array specifier immediately after Task to indicate that the type of the variable is an array of Task type elements.

LiewWS reviewed (commented):
The code is written with strict adherence to the coding standard. Good job!

Commit:
Add ability to delete tasks from list

Duke now has the ability to delete tasks from the list as requested
by the user.

Commit:
Add ability to save tasks to the hard disk

Duke now has the ability to save the list of tasks in the hard disk
automatically whenever the list changes.

Commit:
Merge branch 'branch-Level-6'

Commit:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/duke/Duke.java

Conflicts are resolved.

Commit:
Update Code to edit printed message

One of the print statement in printWelcomeMessage() is modified.

Commit:
Package the App as a JAR file

File created from creating JAR file.

Commit:
Update code to load tasks from file

The code is modified to load and read existing tasks from file,
in addition to saving tasks to file when changed. This change is due
to misunderstanding the requirement of the increment.

Commit:
Update code to improve code quality

Variables in the code are modified to make code better.

jtankw3 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,34 @@
+        return description;
+    }
+
+    public abstract String getTypeOfTask();
```
Since all classes that inherit from `Task` has the same implementation for this method, we can reduce duplicate code by not making this an abstract method

yuen-sihao replied:
Hi, can I clarify if I don't make getTypeOfTask() an abstract method, then what should I initialize the type of task in the Task class as? Is it an empty string?

jtankw3 replied:
Yup that sounds reasonable. You can also consider something like the first answer from this link:
https://stackoverflow.com/questions/2327509/abstract-class-with-final-uninitialized-field 
Both are ok.

jtankw3 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,341 @@
+        tasks.add(newTask);
+    }
+
+    private static File checkIfFileExists(String pathOfDataFile) throws IOException {
```
The name of this method may not be very clear as I would expect it only to check if the file exists but actually it creates a new file.

jtankw3 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,341 @@
+        }
+    }
+
+    private static void determineTypeOfTask(ArrayList<Task> tasks, String typeOfTask, String description) {
```
Similar for this, I would expect it only to return the type of task, but it actually executes the command.

jtankw3 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,341 @@
+    }
+
+    private static void determineTypeOfTask(ArrayList<Task> tasks, String typeOfTask, String description) {
+        switch (typeOfTask) {
```
This variable is a bit confusing as `delete` and `done` are not really tasks types.

yuen-sihao replied:
Have modified typeOfTask to userCommand.

jtankw3 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,341 @@
+        if (tasks.size() > 0) {
+            printListOfTask(tasks);
+        }
+        return pathOfDataFile;
```
Why does `loadListFromFile` return the file path?

yuen-sihao replied:
I realized it is redundant. Have removed it.

jtankw3 reviewed (commented):
Great job, there are almost no coding standard errors! Just take note that some of your method names can be confusing because the name does not capture the actual actions being executed.

Commit:
Refactor code to extract closely related code as classes

Ui, Storage, Parser, and TaskList classes are added.

Commit:
Update code to have user command as individual classes

The user commands are extracted into individual classes.

Commit:
Update code to remove duplicate code

Some parts of the code are duplicated or redundant and are
modified accordingly.

Commit:
Add ability to understand dates

Duke now has the ability to read date inputs.

yuen-sihao committed:
Merge pull request #1 from yuen-sihao/branch-Level-8

Add ability to understand dates

Commit:
Add ability to find a task

Duke is able to allow users to find a task by searching for
a keyword.

Commit:
Add JavaDoc comments

JavaDoc comments are added to the code.

yuen-sihao committed:
Merge pull request #2 from yuen-sihao/branch-Level-9

Add ability to find a task

Commit:
Merge branch 'master' of https://github.com/yuen-sihao/ip

Commit:
Merge branch 'master' into branch-A-JavaDoc

yuen-sihao committed:
Merge pull request #3 from yuen-sihao/branch-A-JavaDoc

Merge branch-A-JavaDoc

Commit:
Update FindCommand with JavaDoc comments

JavaDoc comments are added to the FindCommand class.

Commit:
Update code to understand both dates and time

Duke now can print different formats of dates and time when
required.

Commit:
Modify docs\README.md

The README.md file in the docs folder is modified as a User Guide.

yuen-sihao committed:
Set theme jekyll-theme-architect

yuen-sihao committed:
Set theme jekyll-theme-time-machine

Commit:
Update docs/README.md

Corrected a typo for the releases url and added expected outcome
for welcome screen.

Commit:
Update /docs/README.md

Code is modified to fit the theme better.

Commit:
Update docs/README.md

Edit code due to minor error in heading.
Added new line to docs/_config.yml for standardization.
